lxd/storage/drivers/driver_btrfs_utils.go

```diff
@@ -91,38 +91,39 @@ func (d *btrfs) isSubvolume(path string) bool {
 }
 
 func (d *btrfs) getSubvolumes(path string) ([]string, error) {
-	result := []string{}
+	poolMountPath := GetPoolMountPath(d.name)
+	if !strings.HasPrefix(path, poolMountPath+"/") {
+		return nil, fmt.Errorf("%q is outside pool mount path %q", path, poolMountPath)
+	}
+
+	path = strings.TrimPrefix(path, poolMountPath+"/")
 
 	// Make sure the path has a trailing slash.
 	if !strings.HasSuffix(path, "/") {
 		path = path + "/"
 	}
 
-	// Walk through the entire tree looking for subvolumes.
-	err := filepath.WalkDir(path, func(fpath string, entry fs.DirEntry, err error) error {
-		if err != nil {
-			return err
-		}
+	var stdout bytes.Buffer
+	err := shared.RunCommandWithFds(d.state.ShutdownCtx, nil, &stdout, "btrfs", "subvolume", "list", poolMountPath)
```

Comment:
Do you have an insight into how this command works? I'm wondering if it does anything different than the current implementation in this function does? Does it have access to btrfs internals that makes this operation faster?

Reply:
Ah I just saw:

> It took over three seconds to walk the entire tree in my test environment with a large tree. This can be reduced to a few milliseconds using the command.

Thanks!

Reply:
I suppose it does have a more efficient way to do this then.